POLIO-567 vaccine tab

quang-le opened the pull request:
Add a Vaccine Management tab to the main polio dashboard

## Self proof reading checklist

- [X] Did I use eslint and black formatters
- [X] Is my code clear enough and well documented -  I shall hope so
- [X] Are my typescript files well typed 
- [X] Did I add translations 
- [X] My migrations file are included
- [ ] Are there enough tests - Nope :(
## Changes

### Feature
- The screen is divided in 4 sections: Vaccines, Shipments, Form A and Destruction
- In the vaccines section:
    - The user can add vaccines for a total up to 3 (1 of each type)
    - The user can delete the vaccines one by one from the bottom up
    - The UI will show empty vaccine fields when deleting the last vaccine
    - When the user fills out a vaccine field, the values for reporting delays automatically fill up with default values if they don't have values yet
- In the shipment sections:
    - Same general behavior as vaccines except: 
        - Deleting the last remaining row will remove it from the UI
        - There's no limit to the number of shipment that can be added
        - Filling one field will prompt the validation to force the user to fill all shipment fields or leave them all blank
- Form A and Destruction sections: 
    - If one field is filled up, all related fields but also be filled

### Code
- Front-end
    - Added new components: `VaccineManagementForm`, `RoundVaccineForms` ,`RoundVaccineForm`, `ShipmentForms`, `ShipmentForm` 
    - Updated `useFormValidator` with new fields and new custom validation methods to force user to fill in related fields
- Back-end:
    - Added 2 models: `RoundVaccine` and `Shipment` with a ForeignKey to `Round`
    - Added `create` and `update` methods to `RoundSerializer` to handle the `vaccines` (RoundVaccine) and `shipments`(Shipment) fields


## How to test

- Go to a campaign
- Open the Vaccine Management Tab
- Test for the behavior described above under Feature

## Print screen / video

https://user-images.githubusercontent.com/38907762/191803835-ebba0380-bd2a-48f9-8580-e8aba9a3d56c.mov

## Notes

@olethanh Please have a good look at the `update`method in `RoundSerializer`

quang-le committed:
POLIO-567: WIP outline UI

quang-le committed:
POLIO-567: first draft Vaccine Tab front-end

quang-le committed:
POLIO-567: add translations

quang-le committed:
POLIO-567: add missing translation

quang-le committed:
POLIO-567: add RoundVaccines model

quang-le committed:
POLIO-567: WIP ReportingDelays comp

quang-le committed:
POLIO-567: WIP adapt Vaccine tab to multi vaccine

quang-le committed:
POLIO-567: update serializers for nested fields

quang-le committed:
POLIO-567: update serializer

quang-le committed:
POLIO-567: add transaltions

quang-le committed:
POLIO-567: add Vaccine tab to modal (WIP)

quang-le committed:
POLIO-567: fix API+front-end connection

quang-le committed:
POLIO-567 add form a fields to API

quang-le committed:
POLIO-567 add reporting delays

quang-le committed:
POLIO-567 WIP manage vaccine tabs

quang-le committed:
POLIO-567 remove unused code

quang-le committed:
POLIO-567: update models

quang-le committed:
black formatting

quang-le committed:
Merge branch 'main' into POLIO-567_vaccine_tab

quang-le committed:
POLIO-567: use DecimalField for wastage ratio

quang-le committed:
POLIO-567: add translations

quang-le committed:
POLIO-567: refactor front-end Vaccine mgt

quang-le committed:
POLIO-567: functional API+connect front

quang-le committed:
POLIO-567: send vaccine and shipment id from API

quang-le committed:
black formatting

quang-le committed:
POLIO-567: auto-fill fields with default values

quang-le committed:
POLIO-567 improve layout

quang-le committed:
POLIO-567: update translation to avoid overflow

quang-le committed:
POLIO-567 add translations

quang-le committed:
POLIO-567: add delete btn for vaccine and shipment

quang-le committed:
POLIO-567: allow bigger wastage ratio

quang-le committed:
POLIO-567: add yet more translations

quang-le committed:
POLIO-567: improve UI

- replace icons with buttons
- align new buttons right
- fix bug with auto filled value
- fix some alignment issues

quang-le committed:
POLIO-567: add validation on related fields

quang-le committed:
POLIO-567: another round of translations

quang-le committed:
POLIO-567: improve alignment

quang-le committed:
POLIO-567: cleanup unused code, fix translation

quang-le committed:
POLIO-567: cleanup dead code

quang-le committed:
POLIO-567: allow deleting 1st vaccine

quang-le committed:
POLIO-567 fix python test

beygorghor commented on plugins/polio/js/src/forms/RoundVaccineForm.tsx:
```diff
@@ -0,0 +1,81 @@
+    const accessor = `rounds[${roundIndex}].vaccines[${vaccineIndex}]`;
+
+    return (
+        // <Grid container item xs={12} spacing={2}>
```
you can remove this

beygorghor commented on plugins/polio/js/src/forms/RoundVaccinesForm.tsx:
```diff
@@ -0,0 +1,239 @@
+                })}
+            {vaccines.length === 0 && (
+                <RoundVaccineForm
+                    // eslint-disable-next-line react/no-array-index-key
```
Key is not required here i hink

beygorghor commented on plugins/polio/js/src/forms/VaccineManangementForm.tsx:
```diff
@@ -0,0 +1,190 @@
+                    </Box>
+                </Grid>
+                <Grid container direction="row" item xs={12} spacing={2}>
+                    <Grid item xs={3}>
```
Maybe use a separate component for FormA fields

quang-le replied:
True. I'll make a second pass when updating the number input

beygorghor commented on plugins/polio/js/src/hooks/useFormValidator.js:
```diff
@@ -88,9 +89,257 @@ yup.addMethod(
+    function hasAllFormAFieldsNumber(formatMessage) {
+        return this.test('hasAllFormAFieldsNumber', '', (value, context) => {
+            const { path, createError, parent } = context;
+            const { forma_reception } = parent;
```
`const { forma_reception, forma_missing_vials,  forma_usable_vials, forma_unusable_vials} = parent;` ?

beygorghor commented on plugins/polio/js/src/hooks/useFormValidator.js:
```diff
@@ -88,9 +89,257 @@ yup.addMethod(
+    function hasAllFormAFieldsDate(formatMessage) {
+        return this.test('hasAllFormAFieldsDate', '', (value, context) => {
+            const { path, createError, parent } = context;
+            const { forma_reception } = parent;
```
same here

beygorghor commented on plugins/polio/js/src/hooks/useFormValidator.js:
```diff
@@ -88,9 +89,257 @@ yup.addMethod(
+        return this.test('hasAllShipmentFieldsDate', '', (value, context) => {
+            const { path, createError, parent } = context;
+            const { date_reception } = parent;
+            const { estimated_arrival_date } = parent;
```
same

beygorghor commented on plugins/polio/js/src/hooks/useFormValidator.js:
```diff
@@ -88,9 +89,257 @@ yup.addMethod(
+        return this.test('hasAllShipmentFieldsNumber', '', (value, context) => {
+            const { path, createError, parent } = context;
+            const { date_reception } = parent;
+            const { estimated_arrival_date } = parent;
```
...

beygorghor commented on plugins/polio/js/src/hooks/useFormValidator.js:
```diff
@@ -88,9 +89,257 @@ yup.addMethod(
+        return this.test('hasAllShipmentFieldsString', '', (value, context) => {
+            const { path, createError, parent } = context;
+            const { date_reception } = parent;
+            const { estimated_arrival_date } = parent;
```
...

beygorghor commented on plugins/polio/js/src/pages/Dashboard.js:
```diff
@@ -190,7 +190,7 @@ const Dashboard = ({ router }) => {
                         <>
                             {row.rounds.map(r => (
-                                <li>
+                                <li key={`${r.number}-${r.started_at}`}>
```
This is also fixed on my PR [here](https://github.com/BLSQ/iaso/pull/17)

quang-le replied:
next week ;-)

beygorghor reviewed (approved):
We can improve the view for smaller screens:
![Screenshot 2022-09-23 at 11 54 34](https://user-images.githubusercontent.com/12494624/191936280-54496cff-5890-438d-959a-9d0112447ce7.png)
Also some fields like on print screen pass to error state while only clicking on it.

Nothing blocking for me

olethanh commented on plugins/polio/serializers.py:
```diff
@@ -240,12 +242,91 @@ class Meta:
+    def create(self, validated_data):
+        vaccines = validated_data.pop("vaccines", [])
+        shipments = validated_data.pop("shipments", [])
+        print("VALIDATED DATA", validated_data)
```
print

quang-le committed:
POLIO-567 code review

- adapt layout for smaller (1280) screens
- remove comments and dead code
- refactor validator for clarity

quang-le commented:
> We can improve the view for smaller screens: ![Screenshot 2022-09-23 at 11 54 34](https://user-images.githubusercontent.com/12494624/191936280-54496cff-5890-438d-959a-9d0112447ce7.png) Also some fields like on print screen pass to error state while only clicking on it.
> 
> Nothing blocking for me

I think the thing with the error state comes from the fact that for some parts of the form, all related fields must be filled in together. In your printscreen, you've filled out the vaccine for a shipment, which put all other fields for that shipment in error. When you clicked, `touched` switched to true, so the error was displayed. I can't really think of a workaround at the moment, though it's close to the behaviour we have with rounds

olethanh commented on plugins/polio/serializers.py:
```diff
@@ -240,12 +242,91 @@ class Meta:
+            shipment_instance = shipment_serializer.save()
+            shipment_instances.append(shipment_instance)
+        # remove deleted shipments, ie existing shipments whose id wan't sent in the request
+        all_current_shipments = instance.shipments.all()
```
we delete the shipment but not the vaccine?

olethanh reviewed (approved):
Removed a debug print
It's not very clear when we start filling a shipement line that the Date need to be filled because they never get in red.
It should delete the vaccine line not linked to a round like we does for shipement
And fix the Log modification
But nothing blocking we can do it later
